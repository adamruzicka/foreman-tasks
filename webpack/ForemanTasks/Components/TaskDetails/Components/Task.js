import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Grid, Row, Col, Button } from 'patternfly-react';
import { translate as __ } from 'foremanReact/common/I18n';
import TaskInfo from './TaskInfo';
import { ClickConfirmation } from '../../common/ClickConfirmation';

class Task extends Component {
  taskProgressToggle = () => {
    const {
      timeoutId,
      refetchTaskDetails,
      id,
      loading,
      taskReloadStop,
      taskReloadStart,
    } = this.props;
    if (timeoutId) {
      taskReloadStop(timeoutId);
    } else {
      taskReloadStart(timeoutId, refetchTaskDetails, id, loading);
    }
  };

  render() {
    const {
      taskReload,
      externalId,
      id,
      state,
      allowDangerousActions,
      resumable,
      cancellable,
      hasSubTasks,
      parentTask,
      showUnlockModal,
      showForceUnlockModal,
      toggleUnlockModal,
      toggleForceUnlockModal,
    } = this.props;
    const modalUnlock = (
      <ClickConfirmation
        showModal={showUnlockModal}
        title={__('Unlock')}
        body={__(
          "This will unlock the resources that the task is running against. Please note that this might lead to inconsistent state and should be used with caution, after making sure that the task can't be resumed."
        )}
        confirmationMessage={__(
          'I understand that this may cause harm and have working database backups of all backend services.'
        )}
        confirmAction={__('Unlock')}
        path={`/foreman_tasks/tasks/${id}/unlock`}
        confirmType="warning"
        closeModal={toggleUnlockModal}
      />
    );

    const modalForceUnlock = (
      <ClickConfirmation
        showModal={showForceUnlockModal}
        title={__('Force Unlock')}
        body={__(
          'Resources will be unlocked and will not prevent other tasks from being run. As the task might be still running, it should be avoided to use this unless you are really sure the task got stuck'
        )}
        confirmationMessage={__(
          'I understand that this may cause harm and have working database backups of all backend services.'
        )}
        confirmAction={__('Force Unlock')}
        path={`/foreman_tasks/tasks/${id}/force_unlock`}
        confirmType="danger"
        closeModal={toggleForceUnlockModal}
      />
    );
    return (
      <React.Fragment>
        {modalUnlock}
        {modalForceUnlock}
        <Grid>
          <Row>
            <Col xs={12}>
              <Button
                hidden={!allowDangerousActions}
                className="reload-button"
                bsSize="small"
                onClick={this.taskProgressToggle}
              >
                <span
                  className={`glyphicon glyphicon-refresh ${
                    taskReload ? 'spin' : ''
                  }`}
                />
                {__(`${taskReload ? 'Stop' : 'Start'}  auto-reloading`)}
              </Button>
              <Button
                bsSize="small"
                href={`/foreman_tasks/dynflow/${externalId}`}
              >
                {__('Dynflow console')}
              </Button>
              <Button
                bsSize="small"
                bsStyle="primary"
                data-method="post"
                href={`/foreman_tasks/tasks/${id}/resume`}
                disabled={!resumable}
              >
                {__('Resume')}
              </Button>
              <Button
                bsSize="small"
                data-method="post"
                href={`/foreman_tasks/tasks/${id}/cancel`}
                disabled={!cancellable}
              >
                {__('Cancel')}
              </Button>

              {parentTask && (
                <Button
                  bsSize="small"
                  href={`/foreman_tasks/tasks/${parentTask}`}
                >
                  {__('Parent task')}
                </Button>
              )}
              {hasSubTasks && (
                <Button
                  bsSize="small"
                  href={`/foreman_tasks/tasks/${id}/sub_tasks`}
                >
                  {__('Sub tasks')}
                </Button>
              )}
              {allowDangerousActions && (
                <Button
                  bsSize="small"
                  disabled={state !== 'paused'}
                  onClick={toggleUnlockModal}
                >
                  {__('Unlock')}
                </Button>
              )}
              {allowDangerousActions && (
                <Button
                  bsSize="small"
                  disabled={state === 'stopped'}
                  onClick={toggleForceUnlockModal}
                >
                  {__('Force Unlock')}
                </Button>
              )}
            </Col>
          </Row>
          <TaskInfo {...this.props} />
        </Grid>
      </React.Fragment>
    );
  }
}

Task.propTypes = {
  ...TaskInfo.PropTypes,
  state: PropTypes.string,
  allowDangerousActions: PropTypes.bool,
  resumable: PropTypes.bool,
  cancellable: PropTypes.bool,
  refetchTaskDetails: PropTypes.func,
  hasSubTasks: PropTypes.bool,
  parentTask: PropTypes.string,
  taskReload: PropTypes.bool,
  taskReloadStop: PropTypes.func,
  taskReloadStart: PropTypes.func,
  timeoutId: PropTypes.number,
  externalId: PropTypes.string,
  id: PropTypes.string.isRequired,
  showUnlockModal: PropTypes.bool,
  showForceUnlockModal: PropTypes.bool,
  toggleUnlockModal: PropTypes.func,
  toggleForceUnlockModal: PropTypes.func,
};

Task.defaultProps = {
  ...TaskInfo.defaultProps,
  state: '',
  allowDangerousActions: false,
  resumable: false,
  cancellable: false,
  refetchTaskDetails: () => null,
  hasSubTasks: false,
  parentTask: '',
  taskReload: false,
  taskReloadStop: () => null,
  taskReloadStart: () => null,
  timeoutId: null,
  externalId: '',
  showUnlockModal: false,
  showForceUnlockModal: false,
  toggleUnlockModal: () => null,
  toggleForceUnlockModal: () => null,
};

export default Task;
