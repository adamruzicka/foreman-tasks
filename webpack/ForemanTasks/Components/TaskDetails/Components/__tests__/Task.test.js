import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';

import Task from '../Task';

const fixtures = {
  'render without Props': { id: 'test' },
  'render with some Props': {
    id: 'test',
    state: 'paused',
    hasSubTasks: true,
    allowDangerousActions: true,
  },
};

describe('Task', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(Task, fixtures));
});
