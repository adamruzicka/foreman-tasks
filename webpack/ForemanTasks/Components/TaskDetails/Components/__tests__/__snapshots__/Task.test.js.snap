// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task rendering render with some Props 1`] = `
<Fragment>
  <ClickConfirmation
    body="This will unlock the resources that the task is running against. Please note that this might lead to inconsistent state and should be used with caution, after making sure that the task can't be resumed."
    closeModal={[Function]}
    confirmAction="Unlock"
    confirmType="warning"
    confirmationMessage="I understand that this may cause harm and have working database backups of all backend services."
    path="/foreman_tasks/tasks/test/unlock"
    showModal={false}
    title="Unlock"
  />
  <ClickConfirmation
    body="Resources will be unlocked and will not prevent other tasks from being run. As the task might be still running, it should be avoided to use this unless you are really sure the task got stuck"
    closeModal={[Function]}
    confirmAction="Force Unlock"
    confirmType="danger"
    confirmationMessage="I understand that this may cause harm and have working database backups of all backend services."
    path="/foreman_tasks/tasks/test/force_unlock"
    showModal={false}
    title="Force Unlock"
  />
  <Grid
    bsClass="container"
    componentClass="div"
    fluid={false}
  >
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        componentClass="div"
        xs={12}
      >
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          className="reload-button"
          disabled={false}
          hidden={false}
          onClick={[Function]}
        >
          <span
            className="glyphicon glyphicon-refresh "
          />
          Start  auto-reloading
        </Button>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          disabled={false}
          href="/foreman_tasks/dynflow/"
        >
          Dynflow console
        </Button>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="primary"
          data-method="post"
          disabled={true}
          href="/foreman_tasks/tasks/test/resume"
        >
          Resume
        </Button>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          data-method="post"
          disabled={true}
          href="/foreman_tasks/tasks/test/cancel"
        >
          Cancel
        </Button>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          disabled={false}
          href="/foreman_tasks/tasks/test/sub_tasks"
        >
          Sub tasks
        </Button>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          disabled={false}
          onClick={[Function]}
        >
          Unlock
        </Button>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          disabled={false}
          onClick={[Function]}
        >
          Force Unlock
        </Button>
      </Col>
    </Row>
    <TaskInfo
      action=""
      allowDangerousActions={true}
      cancellable={false}
      endedAt=""
      error={Array []}
      externalId=""
      hasSubTasks={true}
      help=""
      id="test"
      output=""
      parentTask=""
      progress={0}
      refetchTaskDetails={[Function]}
      result="error"
      resumable={false}
      showForceUnlockModal={false}
      showUnlockModal={false}
      startAt=""
      startBefore=""
      startedAt=""
      state="paused"
      taskReload={false}
      taskReloadStart={[Function]}
      taskReloadStop={[Function]}
      timeoutId={null}
      toggleForceUnlockModal={[Function]}
      toggleUnlockModal={[Function]}
      username=""
      usernamePath=""
    />
  </Grid>
</Fragment>
`;

exports[`Task rendering render without Props 1`] = `
<Fragment>
  <ClickConfirmation
    body="This will unlock the resources that the task is running against. Please note that this might lead to inconsistent state and should be used with caution, after making sure that the task can't be resumed."
    closeModal={[Function]}
    confirmAction="Unlock"
    confirmType="warning"
    confirmationMessage="I understand that this may cause harm and have working database backups of all backend services."
    path="/foreman_tasks/tasks/test/unlock"
    showModal={false}
    title="Unlock"
  />
  <ClickConfirmation
    body="Resources will be unlocked and will not prevent other tasks from being run. As the task might be still running, it should be avoided to use this unless you are really sure the task got stuck"
    closeModal={[Function]}
    confirmAction="Force Unlock"
    confirmType="danger"
    confirmationMessage="I understand that this may cause harm and have working database backups of all backend services."
    path="/foreman_tasks/tasks/test/force_unlock"
    showModal={false}
    title="Force Unlock"
  />
  <Grid
    bsClass="container"
    componentClass="div"
    fluid={false}
  >
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        componentClass="div"
        xs={12}
      >
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          className="reload-button"
          disabled={false}
          hidden={true}
          onClick={[Function]}
        >
          <span
            className="glyphicon glyphicon-refresh "
          />
          Start  auto-reloading
        </Button>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          disabled={false}
          href="/foreman_tasks/dynflow/"
        >
          Dynflow console
        </Button>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="primary"
          data-method="post"
          disabled={true}
          href="/foreman_tasks/tasks/test/resume"
        >
          Resume
        </Button>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          data-method="post"
          disabled={true}
          href="/foreman_tasks/tasks/test/cancel"
        >
          Cancel
        </Button>
      </Col>
    </Row>
    <TaskInfo
      action=""
      allowDangerousActions={false}
      cancellable={false}
      endedAt=""
      error={Array []}
      externalId=""
      hasSubTasks={false}
      help=""
      id="test"
      output=""
      parentTask=""
      progress={0}
      refetchTaskDetails={[Function]}
      result="error"
      resumable={false}
      showForceUnlockModal={false}
      showUnlockModal={false}
      startAt=""
      startBefore=""
      startedAt=""
      state=""
      taskReload={false}
      taskReloadStart={[Function]}
      taskReloadStop={[Function]}
      timeoutId={null}
      toggleForceUnlockModal={[Function]}
      toggleUnlockModal={[Function]}
      username=""
      usernamePath=""
    />
  </Grid>
</Fragment>
`;
